File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2018/Y18D07.kt
package org.gristle.pdxpuzzles.advent.y2018

import org.gristle.pdxpuzzles.advent.utilities.Day
import java.util.*

class Y18D07(input: String) : Day {

    // Build two maps. nextSteps shows potential steps that may be assigned once a step is completed
    private val nextSteps: Map<Char, List<Char>> = buildMap<Char, MutableList<Char>> {
        input.lineSequence().forEach { line -> getOrPut(line[5]) { mutableListOf() }.add(line[36]) }
    }

    // precedingSteps shows steps that must be completed before a given step is assigned
    private val precedingSteps: Map<Char, List<Char>> = buildMap<Char, MutableList<Char>> {
        input.lineSequence().forEach { line -> getOrPut(line[36]) { mutableListOf() }.add(line[5]) }
    }

    // starts is the list of steps that have no prerequisites, and thus can be assigned immediately
    private val starts = nextSteps.keys - precedingSteps.keys

    // Worker represents a worker, what it's working on and when it will be ready.
    private data class Worker(var workingOn: Char, var ready: Int) {
        fun isFinished(t: Int) = workingOn.isUpperCase() && ready == t
        fun isIdle() = workingOn == '.'
        fun assign(step: Char, t: Int, offset: Int) {
            workingOn = step
            ready = t + offset + step.code - 64
        }
    }

    // Solve returns both the step sequence and the amount of time required by the workers to compile the sequence.
    // This way the same function can be used for both parts.
    private fun solve(numberOfWorkers: Int, timeOffset: Int): Pair<String, Int> {

        // Use a priority queue to continuously feed the available steps in alphabetical order
        val queue = PriorityQueue<Char>().apply { addAll(starts) }

        // steps tracks the letters that have been delivered
        val steps = mutableSetOf<Char>()

        // numberOfSteps used to terminate the sequence. When steps has all the letters, it will stop.
        val numberOfSteps = (nextSteps.keys + precedingSteps.keys).size

        // represents all the workers available to perform  steps. They begin idle.
        val workerPool = List(numberOfWorkers) { Worker('.', 0) }

        // This sequence starts at second 0 and keeps adding one second.
        // Each second, it harvests completed letters from workers, adding them to the steps
        // It then assigns available letters to idle workers.
        // It terminates when steps contains all the letters, returning the # of seconds.        
        val time = generateSequence(0) { it + 1 }
            .onEach { t -> // for each second...
                // finished workers deliver product and are reset. Newly available steps are added to queue.
                workerPool
                    .filter { it.isFinished(t) } // exclude workers that are not finished
                    .forEach { worker -> // for each finished worker...
                        val product = worker.workingOn
                        steps.add(product) // add finished step to steps
                        worker.workingOn = '.' // reset worker to idle

                        // take the steps that are potentially available now that the worker has completed the step
                        // check that all other preceding steps have already been added, then add to queue
                        val next = nextSteps[product]
                            ?.filter { nextChar -> precedingSteps.getValue(nextChar).all { it in steps } }
                            ?: emptyList()
                        queue.addAll(next)
                    }

                // idle workers are assigned new jobs
                workerPool
                    .filter(Worker::isIdle) // exclude workers that are not idle
                    .forEach { worker -> // for each idle worker...
                        if (queue.isNotEmpty()) { // ...and if a step is available...
                            worker.assign(queue.poll(), t, timeOffset) // ...assign that step to the worker
                        }
                    }
            }.first { steps.size == numberOfSteps } // end sequence and return time once steps is full

        return steps.joinToString("") to time
    }

    override fun part1() = solve(1, 0).first

    override fun part2() = solve(5, 60).second
}

fun main() = Day.runDay(Y18D07::class)

//    Class creation: 12ms
//    Part 1: ABGKCMVWYDEHFOPQUILSTNZRJX (3ms)
//    Part 2: 898 (3ms)
//    Total time: 19ms

File: rust/advent/src/bin/2018/y18d07.rs
use advent::utilities::get_input::get_input;
use indexmap::IndexSet;
use std::{cmp::Reverse, collections::BinaryHeap};
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = (Steps, Steps, Vec<usize>);
type Steps = [Vec<usize>; 26];

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(18, 7).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let mut next_steps: [Vec<usize>; 26] = std::array::from_fn(|_| Vec::new());
    let mut preceding_steps: [Vec<usize>; 26] = std::array::from_fn(|_| Vec::new());
    for line in input.lines() {
        let line = line.as_bytes();
        next_steps[line[5] as usize - 65].push(line[36] as usize - 65);
        preceding_steps[line[36] as usize - 65].push(line[5] as usize - 65);
    }
    let starts = (0..26)
        .filter(|&i| preceding_steps[i].is_empty() && !next_steps.is_empty())
        .collect();
    (next_steps, preceding_steps, starts)
}

#[derive(Clone, Copy, Debug)]
enum WorkerStatus {
    Idle,
    Working(usize),
}

#[derive(Clone, Copy, Debug)]
struct Worker {
    status: WorkerStatus,
    ready: usize,
}

impl Worker {
    fn is_finished(&self, t: usize) -> bool {
        matches!(self.status, WorkerStatus::Working(_)) && self.ready == t
    }

    fn assign(&mut self, step: usize, t: usize, offset: usize) {
        self.status = WorkerStatus::Working(step);
        self.ready = t + offset + step + 1;
    }
}

fn solve(input: &Input, number_of_workers: usize, time_offset: usize) -> (String, usize) {
    let (next_steps, preceding_steps, starts) = input;
    let mut queue: BinaryHeap<Reverse<usize>> =
        starts.iter().map(|&start| Reverse(start)).collect();

    // tracks the letters that have been delivered
    let mut steps = IndexSet::with_capacity(26);

    // represents all the workers available to perform  steps. They begin idle.
    let idle_worker = Worker {
        status: WorkerStatus::Idle,
        ready: 0,
    };
    let mut worker_pool = vec![idle_worker; number_of_workers];

    // This sequence starts at second 0 and keeps adding one second.
    // Each second, it harvests completed letters from workers, adding them to the steps
    // It then assigns available letters to idle workers.
    // It terminates when steps contains all the letters, returning the # of seconds.
    for t in 0.. {
        for worker in worker_pool.iter_mut().filter(|w| w.is_finished(t)) {
            let WorkerStatus::Working(product) = worker.status else {
                panic!("Finished worker has no deliverable");
            };
            steps.insert(product);
            worker.status = WorkerStatus::Idle;

            // take the steps that are potentially available now that the worker has completed the step
            // check that all other preceding steps have already been added, then add to queue
            if let Some(next) = next_steps.get(product) {
                for &next_step in next {
                    if preceding_steps[next_step]
                        .iter()
                        .all(|&it| steps.contains(&it))
                    {
                        queue.push(Reverse(next_step));
                    }
                }
            }
        }

        for worker in worker_pool.iter_mut() {
            if matches!(worker.status, WorkerStatus::Idle) {
                if let Some(Reverse(step)) = queue.pop() {
                    worker.assign(step, t, time_offset);
                } else {
                    break;
                }
            }
        }

        if steps.len() == 26 {
            let letters: String = steps
                .iter()
                .map(|&step| (step as u8 + 65) as char)
                .collect();

            return (letters, t);
        }
    }
    unreachable!()
}

fn part1(input: &Input) -> String {
    solve(input, 1, 0).0
}

fn part2(input: &Input) -> usize {
    solve(input, 5, 60).1
}

#[test]
fn default() {
    let input = get_input(18, 7).unwrap();
    let input = parse_input(&input);
    assert_eq!("ABGKCMVWYDEHFOPQUILSTNZRJX".to_string(), part1(&input));
    assert_eq!(898, part2(&input));
}

// Input parsed (19μs)
// 1. ABGKCMVWYDEHFOPQUILSTNZRJX (31μs)
// 2. 898 (17μs)
// Total: 69μs
